pkg/segment/aggregations/segaggs.go

```diff
@@ -196,6 +209,106 @@ func performAggOnResult(nodeResult *structs.NodeResult, agg *structs.QueryAggreg
 	return nil
 }
 
+func performTail(nodeResult *structs.NodeResult, aggs *structs.QueryAggregators, tailExpr *structs.TailExpr, recs map[string]map[string]interface{}, recordIndexInFinal map[string]int, finishesSegment bool, numTotalSegments uint64, hasSort bool) error {
+
+	if nodeResult.Histogram != nil {
+		for _, aggResult := range nodeResult.Histogram {
+			diff := len(aggResult.Results) - int(aggs.OutputTransforms.TailRows)
+			if diff > 0 {
+				aggResult.Results = aggResult.Results[diff:]
+				aggs.OutputTransforms.TailRows = 0
+			} else {
+				aggs.OutputTransforms.TailRows -= uint64(len(aggResult.Results))
+			}
+			n := len(aggResult.Results)
+			for i := 0; i < n/2; i++ {
+				aggResult.Results[i], aggResult.Results[n-i-1] = aggResult.Results[n-i-1], aggResult.Results[i]
+			}
+			if aggs.OutputTransforms.TailRows == 0 {
+				break
+			}
+		}
+
+		return nil
+	}
+
+	if finishesSegment {
+		tailExpr.NumProcessedSegments++
+	}
+
+	if tailExpr.TailRecords == nil {
+		tailExpr.TailRecords = make(map[string]map[string]interface{}, 0)
+	}
+	if tailExpr.TailPQ == nil {
+		pq := make(utils.PriorityQueue, 0)
+		tailExpr.TailPQ = &pq
+		heap.Init(tailExpr.TailPQ)
+	}
+
+	if !hasSort {
+		for k, v := range recs {
+			timeVal, exists := v["timestamp"]
+			if !exists {
+				continue
+			}
+			heap.Push(tailExpr.TailPQ, &utils.Item{
+				Priority: float64(timeVal.(uint64)),
+				Value:    k,
+			})
+			tailExpr.TailRecords[k] = v
+			if tailExpr.TailPQ.Len() > int(aggs.OutputTransforms.TailRows) {
+				item := heap.Pop(tailExpr.TailPQ).(*utils.Item)
+				delete(tailExpr.TailRecords, item.Value)
+			}
+			delete(recs, k)
+		}
+	}
+
+	if tailExpr.NumProcessedSegments < numTotalSegments {
+		return nil
```

Comment:
I am assuming you are not doing any storing of records like you do if there is no sort...
For a query that has sort, because sort will handle all the segments? So you are just returning if it is not final segment?

Reply:
I think that's correct. But Vasu, you should check if we get an unexpected state (like `sort` is true and this not the last segment but `recs` has some records) and log an error in that case so that we'll be able to detect and debug that quicker if it ever happens

Reply:
You should probably also return an error when you log the error